Sources/src/FromSTMTestBeamData_source.cc

```diff
@@ -0,0 +1,335 @@
+#include "canvas/Persistency/Provenance/ProductID.h"
+#include "canvas/Persistency/Provenance/canonicalProductName.h"
+#include "art/Framework/Core/ModuleMacros.h"
+#include "art/Framework/Principal/Event.h"
+#include "art/Framework/Services/Registry/ServiceHandle.h"
+
+#include "Offline/Sources/inc/STMTestBeamHeaders.hh"
+#include "Offline/DataProducts/inc/STMTestBeamEventInfo.hh"
+#include "Offline/RecoDataProducts/inc/STMWaveformDigi.hh"
+
+
+using namespace std;
+
+namespace mu2e {
+  struct Config
+  {
+    using Name = fhicl::Name;
+    using Comment = fhicl::Comment;
+    fhicl::Sequence<std::string> inputFiles{Name("fileNames"),Comment("Input binary file")};
+    fhicl::Atom<unsigned int> maxEvents{Name("maxEvents"), Comment("Max number of events")};
+    fhicl::Atom<int> verbosityLevel{Name("verbosityLevel"), Comment("Verbosity level")};
+
+    // These are used by art and are required.
+    fhicl::Atom<std::string> module_label{Name("module_label"), Comment("Art module label"), ""};
+    fhicl::Atom<std::string> module_type{Name("module_type"), Comment("Art module type"), ""};
+  };
+  typedef fhicl::WrappedTable<Config> Parameters;
+
+
+  //================================================================
+  class STMTestBeamDataDetail : private boost::noncopyable {
+    std::string myModuleLabel_;
+    art::SourceHelper const& pm_;
+    unsigned runNumber_; // from file name
+    art::SubRunID lastSubRunID_;
+    std::set<art::SubRunID> seenSRIDs_;
+
+    unsigned currentFileNumber_;
+    std::string currentFileName_;
+    std::ifstream* currentFile_ = nullptr;
+
+    unsigned currentSubRunNumber_; // from file
+    unsigned currentEventNumber_;
+    unsigned maxEvents_;
+    int verbosityLevel_;
+    uint16_t channel_; // the channel (HPGe or LaBr) will be different for each file
+    int binaryFileVersion_;
+
+    int printAtEvent;
+
+    // A helper function used to manage the principals.
+    // This is boilerplate that does not change if you change the data products.
+    void managePrincipals ( int runNumber,
+        int subRunNumber,
+        int eventNumber,
+        art::RunPrincipal*&    outR,
+        art::SubRunPrincipal*& outSR,
+        art::EventPrincipal*&  outE);
+
+    public:
+    STMTestBeamDataDetail(const Parameters &conf,
+        art::ProductRegistryHelper &,
+        const art::SourceHelper &);
+
+    void readFile(std::string const& filename, art::FileBlock*& fb);
+
+    bool readNext(art::RunPrincipal* const& inR,
+        art::SubRunPrincipal* const& inSR,
+        art::RunPrincipal*& outR,
+        art::SubRunPrincipal*& outSR,
+        art::EventPrincipal*& outE);
+
+    void closeCurrentFile();
+  };
+
+  //----------------------------------------------------------------
+  STMTestBeamDataDetail::STMTestBeamDataDetail(const Parameters& conf,
+      art::ProductRegistryHelper& rh,
+      const art::SourceHelper& pm)
+    : myModuleLabel_("FromSTMTestBeamData")
```

Comment:
This should come from conf().

Reply:
Done in [51c5470](https://github.com/Mu2e/Offline/pull/856/commits/51c5470641a24465142c7656e928bcb85c0f6b77) and renamed myModuleLabel to myModuleType since the module label is "source" and the module type is "FromSTMTestBeamData"